### ISSUE TITLE
copy! is much slower for Vectors than other Arrays

### ISSUE THREAD

**Issue by npbarnes** [JuliaLang/julia#41985]
I think what's going on is that copy!(::AbstractArray,::AbstractArray) passes it's arguments on to copyto! which is fast for Arrays. (I think copyto! calls the C function memmove. [Here's](https://stackoverflow.com/questions/7776085/why-are-memcpy-and-memmove-faster-than-pointer-increments) a thread that talks about why memmove is fast) However, copy!(::AbstractVector,::AbstractVector) doesn't call copyto!. It does the copy with a simple for loop and the slowdown is substantial. Here's some measurements on my machine. It takes twice as long to copy a 1000000 element Vector than a 1000000 x 1 Matrix even though it's the exact same block of memory.
```julia
julia> a = rand(1000000);
julia> b = similar(a);
julia> @benchmark copy!($b,$a)
BenchmarkTools.Trial: 1939 samples with 1 evaluation.
 Range (min … max):  2.189 ms …   3.961 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     2.486 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   2.544 ms ± 228.841 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

       ▁▃▁▁  ▃▅█▅▆▅▃▂  
  ▁▂▃▆▇████████████████▇▆▅▅▆▆▅▆▃▅▅▅▄▃▃▃▃▄▃▂▃▂▃▂▂▂▃▂▁▂▂▂▂▂▁▁▁▂ ▃
  2.19 ms         Histogram: frequency by time        3.25 ms <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> aa = reshape(a, :, 1);
julia> bb = similar(aa);
julia> @benchmark copy!($bb,$aa)
BenchmarkTools.Trial: 3747 samples with 1 evaluation.
 Range (min … max):  1.105 ms …   2.377 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.292 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.322 ms ± 140.144 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

          ▂▅▅▆▇██▇▇▅▄▃▃▁
  ▂▃▆▇▇▇▇████████████████▇▇▇▄▆▄▅▄▄▃▃▃▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▁▂▁▂▂▁▂▂ ▄
  1.1 ms          Histogram: frequency by time         1.8 ms <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

I think the solution is to replace the for loop in copy!(::AbstractVector,::AbstractVector) with a call to copyto! just like copy!(::AbstractArray,::AbstractArray) defined right below it. Unless there's a good reason to have a for loop there for some AbstractVectors. In that case, copy! should be specialized for Vectors.

**Comment by jishnub** [comment]
Just to add further data, the difference is smaller for me but it exists:

```julia
julia> a = rand(1000000);

julia> b = similar(a);

julia> using BenchmarkTools

julia> @benchmark copy!($b,$a)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     1.371 ms (0.00% GC)
  median time:      1.437 ms (0.00% GC)
  mean time:        1.499 ms (0.00% GC)
  maximum time:     6.252 ms (0.00% GC)
  --------------
  samples:          3283
  evals/sample:     1

julia> aa = reshape(a, :, 1);

julia> bb = similar(aa);

julia> @benchmark copy!($bb,$aa)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     912.552 μs (0.00% GC)
  median time:      972.197 μs (0.00% GC)
  mean time:        1.038 ms (0.00% GC)
  maximum time:     3.229 ms (0.00% GC)
  --------------
  samples:          4713
  evals/sample:     1
```

**Comment by N5N3** [comment]
The relative speed of  `memmove`, `memcopy`, `loop_copy` is heavily dependent the data size you want to copy (at least on my desktop). See [this discourse page](https://discourse.julialang.org/t/use-memcpy-instead-of-memmove-to-copy-array-when-theres-no-overlap/64042/8).

The following benchmark shows that memmove is slower from 128bytes to 2MB.
```julia
julia> function loopcopy!(y,x)
           @inbounds for i ∈ eachindex(y,x)
               y[i] = x[i]
           end
       end
loopcopy! (generic function with 1 method)

julia> memmove!(dest::Array{T}, src::Array{T}, n = length(src)) where T = begin
           t1 = Base.@_gc_preserve_begin dest
           t2 = Base.@_gc_preserve_begin src
           destp = pointer(dest)
           srcp = pointer(src)
           ccall(:memmove, Ptr{Cvoid}, (Ptr{Cvoid}, Ptr{Cvoid}, Csize_t),
               destp, srcp, n * Base.aligned_sizeof(T))
           Base.@_gc_preserve_end t2
           Base.@_gc_preserve_end t1
           return dest
       end
memmove! (generic function with 1 methods)

julia> memcpy!(dest::Array{T}, src::Array{T}, n = length(src)) where T = begin
           t1 = Base.@_gc_preserve_begin dest
           t2 = Base.@_gc_preserve_begin src
           destp = pointer(dest)
           srcp = pointer(src)
           ccall(:memcpy, Ptr{Cvoid}, (Ptr{Cvoid}, Ptr{Cvoid}, Csize_t),
               destp, srcp, n * Base.aligned_sizeof(T))
           Base.@_gc_preserve_end t2
           Base.@_gc_preserve_end t1
           return dest
       end
memcpy! (generic function with 1 methods)

julia> len = 8
8

julia> while len <= 1024 * 1024 * 8
           a = Array{UInt8}(undef,len); b = similar(a);
           println(len,"bytes")
           @btime memcpy!($b, $a)
           @btime memmove!($b, $a)
           @btime loopcopy!($b, $a)
           len *= 2
       end

8bytes
  3.200 ns (0 allocations: 0 bytes)
  6.100 ns (0 allocations: 0 bytes)
  6.600 ns (0 allocations: 0 bytes)
16bytes
  3.200 ns (0 allocations: 0 bytes)
  6.300 ns (0 allocations: 0 bytes)
  7.000 ns (0 allocations: 0 bytes)
32bytes
  2.900 ns (0 allocations: 0 bytes)
  6.100 ns (0 allocations: 0 bytes)
  8.200 ns (0 allocations: 0 bytes)
64bytes
  4.300 ns (0 allocations: 0 bytes)
  8.900 ns (0 allocations: 0 bytes)
  8.600 ns (0 allocations: 0 bytes)
128bytes
  6.100 ns (0 allocations: 0 bytes)
  9.100 ns (0 allocations: 0 bytes)
  7.300 ns (0 allocations: 0 bytes)
256bytes
  14.515 ns (0 allocations: 0 bytes)
  12.212 ns (0 allocations: 0 bytes)
  10.200 ns (0 allocations: 0 bytes)
512bytes
  16.200 ns (0 allocations: 0 bytes)
  15.331 ns (0 allocations: 0 bytes)
  11.712 ns (0 allocations: 0 bytes)
1024bytes
  19.860 ns (0 allocations: 0 bytes)
  22.869 ns (0 allocations: 0 bytes)
  17.034 ns (0 allocations: 0 bytes)
2048bytes
  27.355 ns (0 allocations: 0 bytes)
  37.261 ns (0 allocations: 0 bytes)
  21.142 ns (0 allocations: 0 bytes)
4096bytes
  43.404 ns (0 allocations: 0 bytes)
  66.160 ns (0 allocations: 0 bytes)
  36.455 ns (0 allocations: 0 bytes)
8192bytes
  72.904 ns (0 allocations: 0 bytes)
  124.236 ns (0 allocations: 0 bytes)
  65.102 ns (0 allocations: 0 bytes)
16384bytes
  137.963 ns (0 allocations: 0 bytes)
  289.591 ns (0 allocations: 0 bytes)
  187.097 ns (0 allocations: 0 bytes)
32768bytes
  521.354 ns (0 allocations: 0 bytes)
  727.692 ns (0 allocations: 0 bytes)
  524.339 ns (0 allocations: 0 bytes)
65536bytes
  1.030 μs (0 allocations: 0 bytes)
  1.470 μs (0 allocations: 0 bytes)
  1.080 μs (0 allocations: 0 bytes)
131072bytes
  2.344 μs (0 allocations: 0 bytes)
  3.011 μs (0 allocations: 0 bytes)
  2.433 μs (0 allocations: 0 bytes)
262144bytes
  6.080 μs (0 allocations: 0 bytes)
  7.900 μs (0 allocations: 0 bytes)
  7.500 μs (0 allocations: 0 bytes)
524288bytes
  13.500 μs (0 allocations: 0 bytes)
  17.400 μs (0 allocations: 0 bytes)
  14.800 μs (0 allocations: 0 bytes)
1048576bytes
  27.200 μs (0 allocations: 0 bytes)
  35.400 μs (0 allocations: 0 bytes)
  30.000 μs (0 allocations: 0 bytes)
2097152bytes
  54.700 μs (0 allocations: 0 bytes)
  71.000 μs (0 allocations: 0 bytes)
  60.400 μs (0 allocations: 0 bytes)
4194304bytes
  168.400 μs (0 allocations: 0 bytes)
  147.500 μs (0 allocations: 0 bytes)
  180.200 μs (0 allocations: 0 bytes)
8388608bytes
  597.400 μs (0 allocations: 0 bytes)
  425.400 μs (0 allocations: 0 bytes)
  567.000 μs (0 allocations: 0 bytes)
```

**Comment by fingolfin** [comment]
This seems to be fixed in Julia 1.9 and upwards, based on some quick testing. These are the numbers I get now with Julia 1.9.1 on an M1 Mac (note that `Vector` is slightly faster and has less variance):
```
julia> versioninfo()
Julia Version 1.9.1
Commit 147bdf428cd (2023-06-07 08:27 UTC)
Platform Info:
  OS: macOS (arm64-apple-darwin22.4.0)
  CPU: 10 × Apple M1 Max
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, apple-m1)
  Threads: 1 on 8 virtual cores

julia> using BenchmarkTools

julia> a = rand(1000000);

julia> b = similar(a);

julia> @benchmark copy!($b,$a)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  137.416 μs … 252.125 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     145.209 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   148.484 μs ±   8.910 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

                           ▃█
  ▁▃██▇▇█▇▇▇▇▆▅▄▂▂▁▁▁▁▁▁▁▁▁██▄▂▃▃▂▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  137 μs           Histogram: frequency by time          182 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> aa = reshape(a, :, 1);

julia> bb = similar(aa);

julia> @benchmark copy!($bb,$aa)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  152.708 μs … 358.250 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     156.125 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   157.667 μs ±   8.254 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▂▆▆▇█▇▇█▇▆▄▄▃▂▁▁▁                                             ▂
  ████████████████████▇▇▇▆▆▆▆▆▅▅▃▄▄▄▃▅▆▂▄▅▆▃▇▅▆▆▇▆▇▇▆▇▇▇▆█▇▇▆▆▆ █
  153 μs        Histogram: log(frequency) by time        187 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

```

Here are timings with Julia 1.9.1 on an x86 machine:
```
julia> versioninfo()
Julia Version 1.9.1
Commit 147bdf428cd (2023-06-07 08:27 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 48 × Intel(R) Xeon(R) CPU E5-2687W v4 @ 3.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, broadwell)
  Threads: 1 on 48 virtual cores

julia> using BenchmarkTools

julia> a = rand(1000000);

julia> b = similar(a);

julia> @benchmark copy!($b,$a)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  466.676 μs …  1.835 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     475.425 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   479.618 μs ± 19.679 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

   ▂▆██▆▆▆▄▂▁▁
  ▃█████████████▆▆▅▆▅▄▅▅▄▄▃▄▄▃▃▃▃▃▃▂▃▂▂▂▂▂▂▂▂▂▂▁▂▁▂▁▁▁▂▁▁▁▁▁▁▁ ▃
  467 μs          Histogram: frequency by time          520 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> aa = reshape(a, :, 1);

julia> bb = similar(aa);

julia> @benchmark copy!($bb,$aa)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  466.796 μs …  2.591 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     474.357 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   479.197 μs ± 26.891 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

   ▅█▃
  ▃████▇▇▇▆▅▅▅▄▄▄▃▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  467 μs          Histogram: frequency by time          527 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.

```

**Comment by KristofferC** [comment]
```
julia> @btime copy!($b, $a);
  259.681 μs (0 allocations: 0 bytes)

julia> @btime copy!($bb, $aa);
  259.560 μs (0 allocations: 0 bytes)
```
